flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridgeTest.java

```diff
@@ -0,0 +1,209 @@
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for the {@link DeclarativeSlotPoolBridge}.
+ */
+public class DeclarativeSlotPoolBridgeTest extends TestLogger {
+
+	private static final Time rpcTimeout = Time.seconds(20);
+	private static final JobID jobId = new JobID();
+	private static final JobMasterId jobMasterId = JobMasterId.generate();
+	private final ComponentMainThreadExecutor mainThreadExecutor = ComponentMainThreadExecutorServiceAdapter.forMainThread();
+
+	@Test
+	public void testSlotOffer() throws Exception {
+		final SlotRequestId slotRequestId = new SlotRequestId();
+		final AllocationID expectedAllocationId = new AllocationID();
+		final PhysicalSlot allocatedSlot = createAllocatedSlot(expectedAllocationId);
+
+		final TestingDeclarativeSlotPoolFactory declarativeSlotPoolFactory = new TestingDeclarativeSlotPoolFactory(TestingDeclarativeSlotPool.builder());
+		try (DeclarativeSlotPoolBridge declarativeSlotPoolBridge = createDeclarativeSlotPoolBridge(declarativeSlotPoolFactory)) {
+
+			declarativeSlotPoolBridge.start(jobMasterId, "localhost", mainThreadExecutor);
+
+			CompletableFuture<PhysicalSlot> slotAllocationFuture = declarativeSlotPoolBridge.requestNewAllocatedSlot(slotRequestId, ResourceProfile.UNKNOWN, null);
+
+			declarativeSlotPoolBridge.newSlotsAreAvailable(Collections.singleton(allocatedSlot));
+
+			slotAllocationFuture.join();
+		}
+	}
+
+	@Test
+	public void testNotEnoughResourcesAvailableFailsPendingRequests() throws Exception {
+		final SlotRequestId slotRequestId = new SlotRequestId();
+
+		final TestingDeclarativeSlotPoolFactory declarativeSlotPoolFactory = new TestingDeclarativeSlotPoolFactory(TestingDeclarativeSlotPool.builder());
+		try (DeclarativeSlotPoolBridge declarativeSlotPoolBridge = createDeclarativeSlotPoolBridge(declarativeSlotPoolFactory)) {
+
+			declarativeSlotPoolBridge.start(jobMasterId, "localhost", mainThreadExecutor);
+
+			CompletableFuture<PhysicalSlot> slotAllocationFuture = CompletableFuture
+					.supplyAsync(() -> declarativeSlotPoolBridge.requestNewAllocatedSlot(slotRequestId, ResourceProfile.UNKNOWN, Time.minutes(5)), mainThreadExecutor)
+					.get();
+
+			mainThreadExecutor.execute(() -> declarativeSlotPoolBridge.notifyNotEnoughResourcesAvailable(Collections.emptyList()));
+
+			try {
+				slotAllocationFuture.join();
+				Assert.fail();
+			} catch (Exception e) {
+				Optional<NoResourceAvailableException> expectedException = ExceptionUtils.findThrowable(e, NoResourceAvailableException.class);
+				assertThat(expectedException.isPresent(), is(true));
```

Comment:
`containsCause` is a bit annoying since it also does an exact match against the error message, which tends to just create maintenance overhead down the line.
I will use `futureWillCompleteExceptionally`.